lineapy/graph_reader/utils.py

```diff
@@ -15,3 +20,42 @@ def _is_import_node(graph: Graph, node_id: LineaID) -> bool:
                 lookup_node_name = lookup_node.__dict__.get("name", "")
                 return lookup_node_name in ["l_import", "getattr"]
     return False
+
+
+def get_db_artifacts_from_artifactdef(
+    db: RelationalLineaDB, artifact_entries: List[LineaArtifactDef]
+) -> List[LineaArtifact]:
+    # converts artifactdef list to a linea artfact list by fetching it from the db
+    return [
+        LineaArtifact.get_artifact_from_def(db, art_def)
+        for art_def in artifact_entries
+    ]
+
+
+def get_artifacts_grouped_by_session(
+    all_linea_artifacts: List[LineaArtifact],
+) -> Iterable[Tuple[LineaID, List[LineaArtifact]]]:
+    """
+    Get LineaArtifact from each artifact entry and group by the Session they belong to.
+
+    Artifact entries are specified as name and optionally version as the end user would specify.
+
+    This helper function is used to group target and reuse_precomputed artifacts so that we can
+    create SessionArtifacts for each Session.
+
+    """
+    # This function groups artifacts into a dict with session ids as key
+    #
+    for session_id, artifacts_by_session in groupby(
+        all_linea_artifacts, lambda d: d._session_id
+    ):
+        yield session_id, list(artifacts_by_session)
+
+
```

Comment:
The resulting three functions are simple enough to be unit-testable. They each perform only one function and, hopefully, are equally or more performant.

Reply:
+1